src/bit_str.jl

```diff
@@ -1,127 +1,108 @@
+for IT in [:BigInt, :Int128, :UInt128, :Int64,:UInt64, :Int32, :UInt32, :Int16, :UInt16, :Int8, :UInt8, :Bool]
+    @eval Base.$IT(b::BitStr) = $IT(buf(b))
 end
+for op in [:+, :-, :*, :÷, :|, :⊻, :&, :%, :mod, :mod1]
+    @eval Base.$op(a::T, b::Integer) where T<:BitStr = T($op(buf(a),b))
+    @eval Base.$op(a::Integer, b::T) where T<:BitStr = T($op(a,buf(b)))
+    @eval Base.$op(a::BitStr{N,T}, b::BitStr{N,T}) where {N,T} = BitStr{N,T}($op(buf(a), buf(b)))
+    @eval Base.$op(a::BitStr, b::BitStr) = error("type mismatch: $(typeof(a)), $(typeof(b))")
+end
+Base.:-(x::BitStr{N,T}) where {N,T} = BitStr{N,T}(-buf(x))
 
-BitStr(x::BitStr{T, N}) where {T, N} = BitStr(x.val, N)
-
-"""
-    bit(string)
-
-Create a [`BitStr`](@ref) with given string of bits. See also [`@bit_str`](@ref).
-"""
-bit(x::String) = BitStr(x)
-
-"""
-    bit(x[; len=ndigits(x, base=2)])
-
-Create a [`BitStr`](@ref) accroding to integer `x` to given length `len`.
-"""
-bit(x::Integer; len=ndigits(x, base=2)) = BitStr(x, len)
-
-"""
-    bit(;len)
-
-Lazy curried version of [`bit`](@ref).
-"""
-bit(;len) = x -> bit(x; len=len)
+for op in [:(>>), :(<<)]
+    @eval Base.$op(a::BitStr{N,T}, b::Int) where {N,T} = BitStr{N,T}(Base.$op(buf(a),b))
+    #@eval Base.$op(a::T, b::T) where T<:BitStr = T(Base.$op(buf(a),buf(b)))
+end
 
-"""
-    bit_literal(xs...)
+for op in [:<, :>, :(<=), :(>=)]
+    @eval Base.$op(a::T, b::T) where T<:BitStr = Base.$op(buf(a),buf(b))
+end
 
-Create a [`BitStr`](@ref) by input bits `xs`.
+for op in [:(==)]
+    @eval Base.$op(a::T, b::Number) where T<:BitStr = Base.$op(buf(a),b)
+    @eval Base.$op(a::Number, b::T) where T<:BitStr = Base.$op(a,buf(b))
+    @eval Base.$op(a::BitStr{N}, b::BitStr{N}) where N = Base.$op(buf(a),buf(b))
+end
+for op in [:count_ones, :count_zeros, :leading_ones, :leading_zeros]
+    @eval Base.$op(a::BitStr) = Base.$op(buf(a))
+end
 
-# Example
+# Note: the transitivity of == is not satisfied here.
+Base.:(==)(lhs::BitStr, rhs::BitStr) = false
+Base.isapprox(a::BitStr, b::Number; kwargs...) = Base.isapprox(buf(a),b; kwargs...)
+Base.isapprox(a::Number, b::BitStr; kwargs...) = Base.isapprox(a,buf(b); kwargs...)
+Base.isapprox(lhs::BitStr, rhs::BitStr; kwargs...) = false
+Base.isapprox(a::T, b::T; kwargs...) where T<:BitStr = Base.isapprox(buf(a),buf(b); kwargs...)
 
-```jldoctest
-julia> bit_literal(1, 0, 1, 0, 1, 1)
-110101 (53)
-```
-"""
-bit_literal(xs...) = bit_literal(xs)
-function bit_literal(xs::NTuple{N, Int}) where N
-    val = zero(int(N))
-    for k in 1:N
-        xs[k] == 0 || xs[k] == 1 || error("expect 0 or 1, got $(xs[k])")
-        val += xs[k] << (k - 1)
-    end
-    return BitStr(val, N)
-end
+# Note: it is a bit confusing, with x::BitStr == y::Integer,
+# they behave different when used for indexing.
+Base.to_index(x::BitStr) = error("please do not use bit string for indexing, you may want to use `buf(x)+1` for indexing to avoid ambiguity.")
```

Comment:
we can use `to_location` explicitly, it is convenient and more explicit.

Reply:
We don't have indexing now, so `to_location` Is also not nessesary, right?

Reply:
Oh, I forgot to add deprecate.

Reply:
I think because we don't have `to_index` is the reason to keep this to prevent manual conversion and It's just a convenient one liner anyway.

Reply:
I don't think it provides readability, what is the definition of location? why a bitstring can be converted to a location? According to the code, I guess it is the inverse of to_index, but can not understand why is that.

Reply:
I don't mean inversed `to_index`, I want a one liner function for `x + 1`. Just write

```
to_location(x::Integer) = x + 1
```

would be good and compatible with what I have before. It's not inverse. location has the same definition in circuits, it starts from 1.

Reply:
Fixed, but still in deprecations, `Int64(x) + 1` looks more explicit.